File: examples/calculator-java-layered/src/main/java/com/oopforge/example/layered/calculator/repository/InMemoryCalculationRepository.java
package com.oopforge.example.layered.calculator.repository;

import com.oopforge.example.layered.calculator.domain.Calculation;
import com.oopforge.example.layered.calculator.domain.CalculationId;
import org.springframework.stereotype.Repository;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

@Repository
public class InMemoryCalculationRepository implements CalculationRepository {

    private final Map<CalculationId, Calculation> store = new ConcurrentHashMap<>();

    @Override
    public void save(Calculation calculation) {
        store.put(calculation.id(), calculation);
    }

    @Override
    public Optional<Calculation> findById(CalculationId id) {
        return Optional.ofNullable(store.get(id));
    }
}

File: examples/calculator-python-hexagonal/app/infrastructure/repositories/calculation/in_memory_calculation_repository.py
from app.domain.calculation.model import Calculation
from app.domain.calculation.value import CalculationId


class InMemoryCalculationRepository:
    def __init__(self) -> None:
        self._store: dict[CalculationId, Calculation] = {}

    def save(self, calculation: Calculation) -> None:
        self._store[calculation.id] = calculation

    def find_by_id(self, calculation_id: CalculationId) -> Calculation | None:
        return self._store.get(calculation_id)
